### PR TITLE
Add allen relations

### LINKED ISSUE #995: Add model and problem Allen temporal relations
### Checked for duplicates

Yes - I've already checked

### Alternatives considered

No - I haven't considered

### Related problems

_No response_

### Describe the feature request

Extend the expressiveness of Aerie to represent Allen temporal relations between pairs (timepoint, timepoint), (timepoint, time-interval) or (time-interval, time-interval). These constraints should be able to be applied to the model (e.g intrinsic constrains related to the spacecraft), or problem-dependent

### PR DESCRIPTION
**Pull request by jmdelfa**
* **Tickets addressed:** Closes #995 
* **Review:** By commit  
* **Merge strategy:** Merge (no squash)  

## Description
Implementation of allen temporal relations for coexistence goals. This is achieved by means of a new constraint that allows to define the activity templat start and end timepoints relative to the start and end timepoints of the activity directive

## Verification
Code inspection
Implementation of 7 tests for each of the primary relations in SchedulingIntegrationTests

## Documentation
No documentation has been invalidated.

## Future work
I will produce a new page in the Aerie User's Guide describing how to model these relations, including some examples

### CODE REVIEW

File: scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/models/SchedulingDSL.java [hunk 1]
```diff
@@ -109,20 +120,31 @@ private static final JsonObjectParser<GoalSpecifier.CoexistenceGoalDefinition> c
...
+  coexistenceGoalTransform() {
+    return Convert.between(untuple(GoalSpecifier.CoexistenceGoalDefinition::new), (GoalSpecifier.CoexistenceGoalDefinition $) -> tuple(
+        $.activityTemplate(),
+        $.activityFinder(),
+        $.alias,
+        $.forEach,
+        (Optional<T1>) $.startConstraint,
+        (Optional<T2>) $.endConstraint,
+        $.shouldRollbackIfUnsatisfied
+    ));
```
Comment (adrienmaillard) [inline scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/models/SchedulingDSL.java:145]: Can you give me more details about this ? Why don't we use the usual mapping seen in other parsers in this class ?

File: scheduler-worker/src/test/java/gov/nasa/jpl/aerie/scheduler/worker/services/SchedulingIntegrationTests.java [hunk 2]
```diff
@@ -744,6 +746,445 @@ export default () => Goal.CoexistenceGoal({
...
+
+
+    // Checking that durationParameterActivities finishes at the same time as growBanana
+    final var activitytype = results.plan.getActivitiesByType().keySet().stream().filter(w->w.getName().equals("DurationParameterActivity")).findFirst();
+    if (activitytype.isEmpty())
+      fail("Could not find Coexistence Goal activity type");
+    assertTrue(durationParameterActivity.startOffset().plus(results.plan.getActivitiesByType().get(activitytype.get()).get(0).duration()).noShorterThan(growBanana.startOffset().plus(growBananaDuration).minus(Duration.of(10, Duration.MINUTES))));
+    assertTrue(durationParameterActivity.startOffset().plus(results.plan.getActivitiesByType().get(activitytype.get()).get(0).duration()).noLongerThan(growBanana.startOffset().plus(growBananaDuration).minus(Duration.of(5, Duration.MINUTES))));
+  }
+
```
Comment (adrienmaillard) [inline scheduler-worker/src/test/java/gov/nasa/jpl/aerie/scheduler/worker/services/SchedulingIntegrationTests.java:1259]: Indentation issue ?

File: scheduler-worker/scheduling-dsl-compiler/src/libs/scheduler-edsl-fluent-api.ts [hunk 3]
```diff
@@ -815,6 +839,8 @@ declare global {
      * @param operand the duration offset
      */
     public static range(windowProperty: WindowProperty, operator: TimingConstraintOperator, operand: Temporal.Duration): RangeTimingConstraint
+
+    public static bounds(lowerBound: SingletonTimingConstraint, upperBound: SingletonTimingConstraint): FlexibleRangeTimingConstraint
```
Comment (adrienmaillard) [inline scheduler-worker/scheduling-dsl-compiler/src/libs/scheduler-edsl-fluent-api.ts]: Needs tsdoc

File: scheduler-worker/scheduling-dsl-compiler/src/libs/scheduler-edsl-fluent-api.ts [hunk 4]
```diff
@@ -610,6 +610,14 @@ export class TimingConstraint {
       singleton: false
     })
   }
+
+  public static bounds(lowerBound: SingletonTimingConstraint, upperBound: SingletonTimingConstraint): FlexibleRangeTimingConstraint {
+    return FlexibleRangeTimingConstraint.new({
```
Comment (adrienmaillard) [inline scheduler-worker/scheduling-dsl-compiler/src/libs/scheduler-edsl-fluent-api.ts]: Needs tsdoc

File: scheduler-worker/src/test/java/gov/nasa/jpl/aerie/scheduler/worker/services/MockMerlinService.java [hunk 5]
```diff
@@ -87,7 +88,7 @@ public PlanMetadata getPlanMetadata(final PlanId planId)
         this.missionModelInfo.get().config());
   }
 
-  @Override
+    @Override
```
Comment (adrienmaillard) [inline scheduler-worker/src/test/java/gov/nasa/jpl/aerie/scheduler/worker/services/MockMerlinService.java]: Indent issue ?

**Review by adrienmaillard — CHANGES_REQUESTED** [review]
Looks pretty good to me.

**Comment by camargo** [comment]
Closing for now. Please re-open when we are ready to merge.

**Comment by jmdelfa** [comment]
I discussed with Adrien the status of issue #1087, which might impact the way in which Allen relations are expressed. It seems like #1087 might take some time to be implemented, therefore I will proceed to answer Adrien's comment and close this PR.

File: scheduler-worker/scheduling-dsl-compiler/src/libs/scheduler-edsl-fluent-api.ts (hunk 3 above)
Comment (jmdelfa) [reply scheduler-worker/scheduling-dsl-compiler/src/libs/scheduler-edsl-fluent-api.ts]: Added. You'll see it in the push

File: scheduler-worker/scheduling-dsl-compiler/src/libs/scheduler-edsl-fluent-api.ts (hunk 4 above)
Comment (jmdelfa) [reply scheduler-worker/scheduling-dsl-compiler/src/libs/scheduler-edsl-fluent-api.ts]: Done

File: scheduler-worker/src/test/java/gov/nasa/jpl/aerie/scheduler/worker/services/MockMerlinService.java (hunk 5 above)
Comment (jmdelfa) [reply scheduler-worker/src/test/java/gov/nasa/jpl/aerie/scheduler/worker/services/MockMerlinService.java]: Right. corrected

File: scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/models/SchedulingDSL.java (hunk 1 above)
Comment (jmdelfa) [reply scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/models/SchedulingDSL.java:145]: That's necessary to convert from VariadicProductParser to JsonObjectParser. After speaking to Matt (with whom I pair programmed the code of this PR) and making some further investigation, I don't see any other parser method that could do this same job.

**Comment by jmdelfa** [comment]
Commit includes comments and indentation modifications requested by @adrienmaillard

**Review by adrienmaillard — APPROVED** [review]
LGTM

**Pull request merged**